### ISSUE TITLE
Performance issue in Serializer when using dynamic schema registry with avro schema

### ISSUE THREAD

**Issue by TomBillietKlarrio** [Apicurio/apicurio-registry#2483]
We're using apicurio with a dynamic avro schema (we don't know it upfront), so we have dynamic schema registry enabled. Functionality it works, but we're seeing a heavy CPU usage on the avro serializer.

I did some profiling and digging into the code and came to the following conclusion:
Most of the CPU cycles are spent in the `schema.toString()` call here:
https://github.com/Apicurio/apicurio-registry/blob/cbf08dd1e16b00c8c8532110a91b49879ae87c50/serdes/avro-serde/src/main/java/io/apicurio/registry/serde/avro/AvroSchemaParser.java#L75
This is called from https://github.com/Apicurio/apicurio-registry/blob/cbf08dd1e16b00c8c8532110a91b49879ae87c50/schema-resolver/src/main/java/io/apicurio/registry/resolver/DefaultSchemaResolver.java#L80

However, a few lines below there is:
```
if(schemaCache.containsByArtifactReference(artifactReference)) {
    return resolveSchemaByArtifactReferenceCached(artifactReference);
}
```

Now there are 2 problems I see here:
* For doing this lookup in the cache, there's not need for the expensive `schema.toString()` call to be done. It doesn't need the string representation to lookup in the cache. So it's wasteful that it's done already.
* I always see a cache miss. I haven't been able to pinpoint exactly why, but it looks to me that because of the dynamic registration feature the entries in the cache always seem to have a `globalId` set, while the entry that's being compared has this field set to `null`, so it will never match.

At the moment we worked around this issue by extending the `DefaultSchemaResolver` like this:
```java
public class CachedSchemaResolver<S, T> extends DefaultSchemaResolver<S, T> {

    private static final Logger LOG =
            LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
    
    // this cache is a slow memory leak, we only add to it
    private final Map<Schema, SchemaLookupResult<S>> cache = new ConcurrentHashMap<>();
    private final AvroDatumProvider<T> avroDatumProvider = new DefaultAvroDatumProvider<>();

    @Override
    public SchemaLookupResult<S> resolveSchema(Record<T> data) {
        Objects.requireNonNull(data);
        Objects.requireNonNull(data.payload());

        // Make sure there is no schema.toString() call somewhere in this codepath. It causes a lot of cpu cycles.
        Schema schema = avroDatumProvider.toSchema(data.payload());
        return cache.computeIfAbsent(schema, s -> {
            LOG.debug("Cache miss for {}", data.metadata().artifactReference());
            return super.resolveSchema(data);
        });
    }
}
```

It gives us a 3 times lower CPU usage as with the default implementation.

I'm also willing to work on a fix if you could give me some pointers on why this causes a cache miss every time.

**Comment by carlesarnal** [comment]
Hi @TomBillietKlarrio,

The fix I'm working on to fix the #2283 issue is going to solve this issue, I think since it's going to solve the cache issues.

Thanks!

**Comment by TomBillietKlarrio** [comment]
ok thanks for the update!
